src/frontend/src/optimizer/plan_node/stream_group_topn.rs

```diff
@@ -52,10 +52,18 @@ impl StreamGroupTopN {
             watermark_columns
         };
 
+        // We can use the group key as the stream key when limit is 1, since there is only one
+        // record for each value of group key.
+        let logical_pk = if logical.limit == 1 {
```

Comment:
We may also check `with_ties` here to avoid deriving an incorrect primary key for `RANK()`.